pyiron_base/interfaces/lockable.py

```diff
@@ -0,0 +1,347 @@
+
+    >>> class MyLock(Lockable):
+    ...   def __init__(self, foo):
+    ...     super().__init__()
+    ...     self._foo = foo
+    ...   @property
+    ...   def foo(self):
+    ...     return self._foo
+    ...   @foo.setter
+    ...   @sentinel
+    ...   def foo(self, value):
+    ...     self._foo = value
+    >>> ml = MyLock(42)
+    >>> ml.foo
+    42
+    >>> ml.foo = 23
+    >>> ml.lock()
+    >>> ml.foo = 42
+    Traceback (most recent call last):
+        ...
+    lockable.Locked: Object is currently locked!  Use unlocked() if you know what you are doing.
+
+    It's possible to change the errors raised into a warning and allow
+    modification by passing `lock_method` to :meth:`~.Lockable.__init__` or
+    `method` to :meth:`~.lock`.
+
+    >>> mw = LockList(lock_method="warning")
+    >>> mw.append(0)
+    >>> mw.lock()
+    >>> mw[0] = 1 # will print the warning
+    >>> mw[0]
+    1
+
+    >>> mw = LockList()
+    >>> mw.append(0)
+    >>> mw.lock(method='warning')
+    >>> mw[0] = 1 # will print the warning
+    >>> mw[0]
+    1
+
+    """
+
+    def __init__(self, *args, lock_method: str = "error", **kwargs):
+        object.__setattr__(self, "_read_only", False)
+        object.__setattr__(self, "_lock_method", lock_method)
+        super().__init__(*args, **kwargs)
+
+    @property
+    def read_only(self) -> bool:
+        """
+        bool: False if the object can currently be written to
+
+        Setting this value will trigger :meth:`._on_lock` and :meth:`._on_unlock` if it changes.
+        """
+        return object.__getattribute__(self, "_read_only")
+
+    @read_only.setter
+    def read_only(self, value: bool):
+        changed = self._read_only != value
+        if changed:
+            self._read_only = value
+            if value:
+                self._on_lock()
+            else:
+                self._on_unlock()
+
+    def _on_lock(self):
+        for it in _iterate_lockable_subs(self):
+            it.lock()
+
+    def _on_unlock(self):
+        for it in _iterate_lockable_subs(self):
+            it.read_only = False
+
+    def lock(self, method: Optional[Literal["error", "warning"]] = None):
+        """
+        Set :attr:`~.read_only`.
+
+        Objects may be safely locked multiple times without further effect.
+
+        Args:
+            method (str, either "error" or "warning"): if "error" raise an :class:`.Locked` exception if modification is
+                    attempted; if "warning" raise a :class:`.LockedWarning` warning; default is "error" or the value
+                    passed to the constructor.
+
+        Raises:
+            ValueError: if `method` is not an allowed value
+        """
+        if method not in ["error", "warning", None]:
+            raise ValueError(f"Unrecognized lock method {method}!")
+        if method is not None:
+            object.__setattr__(self, "_lock_method", method)
+        self.read_only = True
+
+    def unlocked(self) -> _UnlockContext:
```

Comment:
nit: I think this should be `unlock` to match the verb tense of `lock`

Reply:
I named it asymmetrical on purpose, because `lock` and `unlocked` are not symmetric either.

Reply:
What I wanted to avoid is that some code unlocks an object and then crashes (or forgets it) before it can relock it.  So `obj.unlocked()` is a context manager that does the relocking.

Reply:
Clever. I approve. But I was too dense to pick up on it on the first read, so maybe a comment or docstring to clarify intent is good.